src/statsmodels/contrasts.jl

```diff
@@ -71,49 +69,50 @@ contrasts_matrix(C::MyContrasts, baseind, n) = ...
 """
 abstract AbstractContrasts
 
-## Contrasts + Data = ContrastsMatrix
-type ContrastsMatrix{T}
+# Contrasts + Levels (usually from data) = ContrastsMatrix
+type ContrastsMatrix{T, S <: AbstractContrasts}
     matrix::Matrix{Float64}
     termnames::Vector{T}
     levels::Vector{T}
-    contrasts::AbstractContrasts
+    contrasts::S
 end
 
 """
-    ContrastsMatrix{T}(::AbstractContrasts, levels::Vector{T})
+    ContrastsMatrix{T}(C::AbstractContrasts, levels::Vector{T})
 
-Instantiate contrasts matrix for given data (categorical levels)
+Compute contrasts matrix for a given set of categorical data levels.
 
 If levels are specified in the `AbstractContrasts`, those will be used, and likewise
 for the base level (which defaults to the first level).
 """
-function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
-
-    ## if levels are defined on C, use those, validating that they line up.
-    ## what does that mean?
-    ##
-    ## C.levels == lvls (best case)
-    ## data levels missing from contrast: would generate empty/undefined rows. 
-    ## better to filter data frame first
-    ## contrast levels missing from data: would have empty columns, generate a
-    ## rank-deficient model matrix.
-    c_lvls = convert(typeof(lvls), get(C.levels, lvls))
-    mismatched_lvls = symdiff(c_lvls, lvls)
-    isempty(mismatched_lvls) || error("Contrasts levels not found in data or vice-versa: ", mismatched_lvls)
-
-    n = length(c_lvls)
-    n > 1 || error("not enough degrees of freedom to define contrasts")
-    
-    ## find index of base level. use C.base, then default (1).
-    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
-    baseind > 0 || error("Base level $(C.base) not found in levels")
+function ContrastsMatrix{T}(C::AbstractContrasts, levels::Vector{T})
+
+    # if levels are defined on C, use those, validating that they line up.
+    # what does that mean? either:
+    #
+    # 1. C.levels == levels (best case)
+    # 2. data levels missing from contrast: would generate empty/undefined rows. 
+    #    better to filter data frame first
+    # 3. contrast levels missing from data: would have empty columns, generate a
+    #    rank-deficient model matrix.
+    c_levels = oftype(levels, get(C.levels, levels))
+    mismatched_levels = symdiff(c_levels, levels)
+    isempty(mismatched_levels) || error("contrasts levels not found in data or vice-versa: $mismatched_levels.\nData levels: $levels.\nContrast levels: $c_levels")
+
+    n = length(c_levels)
+    n == 0 && error("empty set of levels found (need at least two to compute contrasts).")
+    n == 1 && error("only one level found: $(c_levels[1]). need at least two to compute contrasts.")
```

Comment:
Upper case "N". There's an inconsistency in the punctuation w.r.t. the message above.